src/stdio/src/lib.rs

```diff
@@ -1,41 +1,186 @@
+        let mut advance = 0;
+        let mut f_buf: &'static _ = unsafe { slice::from_raw_parts(self.wbase, len) };
+        let mut f_filled = false;
+        let mut rem = f_buf.len() + to_write.len();
+        loop {
+            let mut count = if f_filled {
+                platform::write(self.fd, &f_buf[advance..])
+            } else {
+                platform::write(self.fd, &f_buf[advance..]) + platform::write(self.fd, to_write)
+            };
+            if count == rem as isize {
+                self.wend = unsafe { self.buf.add(self.buf_size - 1) };
+                self.wpos = self.buf;
+                self.wbase = self.buf;
+                return to_write.len();
+            }
+            if count < 0 {
+                self.wpos = ptr::null_mut();
+                self.wbase = ptr::null_mut();
+                self.wend = ptr::null_mut();
+                self.flags |= constants::F_ERR;
+                return 0;
+            }
+            rem -= count as usize;
+            if count as usize > len {
+                count -= len as isize;
+                f_buf = unsafe { mem::transmute(to_write) };
+                f_filled = true;
+                advance = 0;
+            }
+            advance += count as usize;
+        }
+    }
+    pub fn read(&mut self, buf: &mut [u8]) -> usize {
+        use core::slice;
+        // let buff = slice::from_raw_parts_mut(buf, size - !((*stream).buf_size == 0) as usize);
+        let adj = !(self.buf_size == 0) as usize;
+        let mut file_buf: &'static mut _ =
+            unsafe { slice::from_raw_parts_mut(self.buf, self.buf_size) };
+        let count = if buf.len() <= 1 + adj {
+            platform::read(self.fd, file_buf)
+        } else {
+            platform::read(self.fd, buf) + platform::read(self.fd, file_buf)
+        };
+        if count <= 0 {
+            self.flags |= if count == 0 {
+                constants::F_EOF
+            } else {
+                constants::F_ERR
+            };
+            return 0;
+        }
+        if count as usize <= buf.len() - adj {
+            return count as usize;
+        }
+        unsafe {
+            // Adjust pointers
+            self.rpos = self.buf;
+            self.rend = self.buf.offset(count);
+            buf[buf.len() - 1] = *self.rpos;
+            self.rpos = self.rpos.add(1);
+        }
+        buf.len()
+    }
+    pub fn seek(&self, off: off_t, whence: c_int) -> off_t {
+        unsafe { platform::lseek(self.fd, off, whence) }
+    }
+}
+impl Write for FILE {
+    fn write_str(&mut self, s: &str) -> Result {
+        let s = s.as_bytes();
+        if self.write(s) != s.len() {
+            Err(Error)
+        } else {
+            Ok(())
+        }
+    }
+}
 
+/// Clears EOF and ERR indicators on a stream
 #[no_mangle]
-pub extern "C" fn clearerr(stream: *mut FILE) {
-    unimplemented!();
+pub extern "C" fn clearerr(stream: &mut FILE) {
+    stream.flags &= !(F_EOF & F_ERR);
```

Comment:
I think this should be
```rust
stream.flags &= !(F_EOF | F_ERR);
```
Otherwise I believe it'll just do nothing.

Reply:
Ah, good catch